boot/fdehook/fdehook.go

```diff
@@ -0,0 +1,131 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package fdehook
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	"github.com/snapcore/snapd/osutil"
+)
+
+func fdeHook(kernelDir string) string {
+	return filepath.Join(kernelDir, "meta/hooks/fde")
+}
+
+// Enabled returns whether the external FDE helper should be called
+func Enabled(kernelDir string) bool {
+	return osutil.FileExists(fdeHook(kernelDir))
+}
+
+// fdeHookCmd returns a *exec.Cmd that runs the fdehook code with
+// (some) sandboxing around it. The sandbox will ensure that it's hard
+// for the hook to abuse that they are called in e.g. initrd.  But it
+// does not aim for perfect protection - the fdehooks are part of the
+// kernel snap/initrd so if someone wants to do mischief there are
+// easier ways by hacking initrd directly.
+func fdeHookCmd(kernelDir string, args ...string) *exec.Cmd {
+	cmd := exec.Command(
+		"systemd-run",
+		append([]string{
+			"--pipe", "--same-dir", "--wait", "--collect",
+			"--service-type=exec", "--quiet",
+			// do not allow mounting, this ensures hooks
+			// in initrd can not mess around with
+			// ubuntu-data
+			"--property=SystemCallFilter=~@mount",
+			// add basic sandboxing to prevent messing
+			// around
+			// XXX: this maybe too strict, i.e. to unseal
+			//      the fdehook may need to write to some
+			//      crypto device in /dev which will not be
+			//      possible with "ProtectSystem=strict"
+			"--property=ProtectSystem=strict",
+			fdeHook(kernelDir),
+		}, args...)...)
```

Comment:
I would think about adding a timeout, so that the hook must unlock the partition in a reasonable amount of time or have the boot fail with a reasonable message.

Reply:
Timeout added